bumble/profiles/asha_service.py

```diff
@@ -0,0 +1,138 @@
+  UUID = GATT_ASHA_SERVICE
+
+  def __init__(self,device:Device):
+    self.device=device
+
+    # Handler for volume control
+    def on_volume_write(connection, value):
+      print('--- VOLUME Write:', value[0])
+
+    # Handler for audio control commands
+    def on_audio_control_point_write(connection, value):
+      print('--- AUDIO CONTROL POINT Write:', value.hex())
+      opcode = value[0]
+      if opcode == 1:
+        # Start
+        audio_type = ('Unknown', 'Ringtone', 'Phone Call', 'Media')[value[2]]
+        print(
+            f'### START: codec={value[1]}, audio_type={audio_type}, volume={value[3]}, otherstate={value[4]}')
+      elif opcode == 2:
+        print('### STOP')
+      elif opcode == 3:
+        print(f'### STATUS: connected={value[1]}')
+
+      # TODO Respond with a status
+      # asyncio.create_task(device.notify_subscribers(audio_status_characteristic, force=True))
+
+    self.read_only_properties_characteristic = Characteristic(
+        GATT_ASHA_READ_ONLY_PROPERTIES_CHARACTERISTIC,
+        Characteristic.READ,
+        Characteristic.READABLE,
+        bytes([
+            0x01,  # Version
+            0x00,  # Device Capabilities [Left, Monaural]
+            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,  # HiSyncId
+            0x01,  # Feature Map [LE CoC audio output streaming supported]
+            0x00, 0x00,  # Render Delay
+            0x00, 0x00,  # RFU
+            0x02, 0x00  # Codec IDs [G.722 at 16 kHz]
+        ])
+    )
+
+    self.audio_control_point_characteristic = Characteristic(
+        GATT_ASHA_AUDIO_CONTROL_POINT_CHARACTERISTIC,
+        Characteristic.WRITE | Characteristic.WRITE_WITHOUT_RESPONSE,
+        Characteristic.WRITEABLE,
+        CharacteristicValue(write=on_audio_control_point_write)
+    )
+    self.audio_status_characteristic = Characteristic(
+        GATT_ASHA_AUDIO_STATUS_CHARACTERISTIC,
+        Characteristic.READ | Characteristic.NOTIFY,
+        Characteristic.READABLE,
+        bytes([0])
+    )
+    self.volume_characteristic = Characteristic(
+        GATT_ASHA_VOLUME_CHARACTERISTIC,
+        Characteristic.WRITE_WITHOUT_RESPONSE,
+        Characteristic.WRITEABLE,
+        CharacteristicValue(write=on_volume_write)
+    )
+
+    # TODO add real psm value
+    self.psm=0x0080
+    # self.psm = device.register_l2cap_channel_server(0, on_coc, 8)
+    self.le_psm_out_characteristic = Characteristic(
+        GATT_ASHA_LE_PSM_OUT_CHARACTERISTIC,
+        Characteristic.READ,
+        Characteristic.READABLE,
+        struct.pack('<H', self.psm)
+    )
+
+    characteristics = [self.read_only_properties_characteristic,
+                       self.audio_control_point_characteristic,
+                       self.audio_status_characteristic,
+                       self.volume_characteristic,
+                       self.le_psm_out_characteristic]
+
+    super().__init__(characteristics)
+
+  async def start_advertising(self, capability: int,
```

Comment:
rather than duplicating the `start_advertising` method of `Device`, I would recommend just creating a factory method that can construct and return the advertising data, and let the client then combine it with its own data (like name, maybe other services) and then start advertising. Only the `SERVICE_DATA_16_BIT_UUID` is really needed in fact, since the `INCOMPLETE_LIST_OF_16_BIT_SERVICE_CLASS_UUIDS` would also be something the client may want to add to.

Reply:
ack